### DISCUSSION TITLE
Switch Avatto 4 gangs with socket unsupported

### DISCUSSION THREAD

**Discussion by cmagno369** [Koenkk/zigbee2mqtt#20014] [Devices]
I bought the avatto brand switch, TS0006 which is not yet supported by zigbee2mqtt. I read that it is possible to create an external conversion to work with the model.
I followed the step by step for models with more than 1 gang but it activates all gangs at once. Now it is showing an error when executing the code from index.js.
Are there any similar external conversion models that work?

Code:

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['TS0006'],
    model: '_TZ3000_cvis4qmw',
    vendor: 'AVATTO',
    description: '4x4 Smart Touch Switch whith Socket',    
    fromZigbee: [fz.ignore_basic_report, fz.tuya_switch,fz.on_off],
    toZigbee: [tz.tuya_switch_state,tz.on_off],
    meta: {multiEndpoint: true},
    exposes: [e.switch().withEndpoint('l1').setAccess('state', ea.STATE_SET),
        e.switch().withEndpoint('l2').setAccess('state', ea.STATE_SET),
        e.switch().withEndpoint('l3').setAccess('state', ea.STATE_SET),
        e.switch().withEndpoint('l4').setAccess('state', ea.STATE_SET),
        e.switch().withEndpoint('l5').setAccess('state', ea.STATE_SET),
        e.switch().withEndpoint('l6').setAccess('state', ea.STATE_SET)],
    whiteLabel: [
        {vendor: 'Norklmes', model: 'MKS-CM-W5'},
        {vendor: 'Somgoms', model: 'ZSQB-SMB-ZB'},
        {vendor: 'Moes', model: 'WS-EUB1-ZG'},
        {vendor: 'AVATTO', model: 'ZGB-WS-EU'},
    ],
    endpoint: (device) => {
        // Endpoint selection is made in tuya_switch_state
        return {'l1': 1, 'l2': 1, 'l3': 1, 'l4': 1, 'l5': 1, 'l6': 1};
    }
};
module.exports = definition;

**Comment by Koenkk** [comment]
Try adding `configure: tuya.configureMagicPacket,` below `meta`

**Comment by cmagno369** [comment]
Thanks for your help.
I added the configure suggested but the Zig2M respond with the error below:

/app/node_modules/zigbee-herdsman-converters/index.js:172
        if (converter.options) {
                      ^
TypeError: Cannot read properties of undefined (reading 'options')
    at Object.addDefinition [as addDeviceDefinition] (/app/node_modules/zigbee-herdsman-converters/index.js:172:23)

I include the lib tuya:
const tuya = require('zigbee-herdsman-converters/lib/tuya');

and added like you mentioned:
    meta: {multiEndpoint: true},
    configure: tuya.configureMagicPacket,

**Reply by Koenkk** [reply]
Add the following to the top of the file `const tuya = require('zigbee-herdsman-converters/lib/tuya');`

**Comment by cmagno369** [comment]
Hi @Koenkk 
Thanks for the reply.
I entered the code as indicated, the error did not occur, but the switch still appears as unsupported. I tried another approach without success as well.
Follow my new approach:

const tuya = require('zigbee-herdsman-converters/lib/tuya');
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
     fingerprint: [{modelID: 'TS0006', manufacturerName: '_TZ3000_cvis4qmw'}],
        model: 'TS0006',
        vendor: 'TuYa',
        description: '4 gang switch module with neutral wire and socket',
        extend: tuya.extend.switch(),
        exposes: [e.switch().withEndpoint('l1'), e.switch().withEndpoint('l2'), e.switch().withEndpoint('l3'),
            e.switch().withEndpoint('l4'), e.switch().withEndpoint('l5'), e.switch().withEndpoint('l6')],
        endpoint: (device) => {
            return {'l1': 1, 'l2': 2, 'l3': 3, 'l4': 4, 'l5': 5, 'l6': 6};
        },
        meta: {multiEndpoint: true},
        configure: async (device, coordinatorEndpoint, logger) => {
            await tuya.configureMagicPacket(device, coordinatorEndpoint, logger);
            await reporting.bind(device.getEndpoint(1), coordinatorEndpoint, ['genOnOff']);
            await reporting.bind(device.getEndpoint(2), coordinatorEndpoint, ['genOnOff']);
            await reporting.bind(device.getEndpoint(3), coordinatorEndpoint, ['genOnOff']);
            await reporting.bind(device.getEndpoint(4), coordinatorEndpoint, ['genOnOff']);
            await reporting.bind(device.getEndpoint(5), coordinatorEndpoint, ['genOnOff']);
            await reporting.bind(device.getEndpoint(6), coordinatorEndpoint, ['genOnOff']);
            await reporting.onOff(endpoint);
        },
};

module.exports = definition;

**Comment by cmagno369** [comment]
I review my entire approach and make some changes on the code. 
For now Zigbee2mqtt recognize the switch and expose the endpoint. 
I'll do some more tests, but for now it's works fine.
Follow the entire code:

const tuya = require('zigbee-herdsman-converters/lib/tuya');
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    fingerprint: tuya.fingerprint('TS0006', ['_TZ3000_cvis4qmw']),
        model: 'TS0006',
        vendor: 'AVATTO',
        description: '4 gang switch module with neutral wire and socket',
        extend: tuya.extend.switch({powerOnBehavior2: true, indicatorMode: true, backlightModeOffOn: true, endpoints: ['l1', 'l2', 'l3','l4','l5','l6']}),
        endpoint: (device) => {
            return {'l1': 1, 'l2': 2, 'l3': 3, 'l4': 4, 'l5': 5, 'l6': 6};
        },
        meta: {multiEndpoint: true},
        configure: async (device, coordinatorEndpoint, logger) => {
            await tuya.configureMagicPacket(device, coordinatorEndpoint, logger);
            await reporting.bind(device.getEndpoint(1), coordinatorEndpoint, ['genOnOff']);
            await reporting.bind(device.getEndpoint(2), coordinatorEndpoint, ['genOnOff']);
            await reporting.bind(device.getEndpoint(3), coordinatorEndpoint, ['genOnOff']);
            await reporting.bind(device.getEndpoint(4), coordinatorEndpoint, ['genOnOff']);
            await reporting.bind(device.getEndpoint(5), coordinatorEndpoint, ['genOnOff']);
            await reporting.bind(device.getEndpoint(6), coordinatorEndpoint, ['genOnOff']);
        },
        whiteLabel: [
            tuya.whitelabel('AVATTO', 'TS0006', '4 gang switch module with neutral wire and socket', ['_TZ3000_cvis4qmw']),
        ],
};

module.exports = definition;

**Reply by Koenkk** [reply]
Integrated the changes!

Changes will be available in the [dev branch](https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html) in a few hours from now.
